graph.go

```diff
@@ -71,41 +77,209 @@ func Graph(units unit.SourceUnits) (*graph.Output, error) {
+type addSelector func(s selector)
+type selectorExists func(s selector) bool
+
+var lastSelectorRegexp = regexp.MustCompile(".*([\\.\\#].+)")
+
+func selSplitFn(r rune) bool {
+	return r == '>' || r == '+' || r == '~'
+}
+
+// lastSelector returns the last selector from given selector chain.
+func lastSelector(s string) *selector {
+	// Splits `s` into a slice of selectors. It might be a single selector or a chain of selectors(Eg. ".panel > .panel-body + .table").
+	selectors := strings.FieldsFunc(s, selSplitFn)
+
+	// sel is the single or last selector from selectors chain.
+	sel := strings.TrimSpace(selectors[len(selectors)-1])
+
+	// sel might still be a chain of selectors(Eg. "h1.title")
+	// `lastSelectorRegexp` is used to obtain the last selector which starts either with "." or "#".
+	m := lastSelectorRegexp.FindStringSubmatch(sel)
+	if len(m) != 2 {
+		return nil
+	}
+	var l selector
+	l = selector(m[1])
+	return &l
+}
+
+func doGraph(u *unit.SourceUnit) (*graph.Output, error) {
+	out := graph.Output{}
+
+	// At the moment we support a unique selector per `out`.
+	var (
+		// sels are used to keep track which selectors exists for `out`.
+		sels selectors = make(selectors, 0)
+
+		// selExists is used to check if a given selector exits in `sels`.
+		selExists selectorExists
+
+		// addSel is used to add a given selector to `sels`.
+		addSel addSelector
+	)
+	selExists = func(s selector) bool {
+		if _, found := sels[s]; found {
+			return true
+		}
+		return false
+	}
+	addSel = func(s selector) {
+		sels[s] = s
+	}
 
 	// Iterate over u.Files, for each file the process performed can be described as follow:
-	// 1. The file is CSS parsed.
-	// 2. For each CSS property found on the parsed file a graph.Ref is created & appended to out.Refs.
-	for _, currentFile := range u.Files {
-		f, err := ioutil.ReadFile(currentFile)
+	// 1. Read the file and parse its data.
+	// 2. For CSS files: for each selector found a graph.Def is created and for each property of that selector a graph.Ref is created.
+	// 3. For HTML files: for each tag attribute(id and class) a graph.Ref is created.
+	for _, f := range u.Files {
+		fileBytes, err := ioutil.ReadFile(f)
 		if err != nil {
 			log.Printf("failed to read a source unit file: %s", err)
 			continue
 		}
-		file := string(f)
-		stylesheet, err := cssParser.Parse(file)
-		if err != nil {
-			return nil, err
-		}
-		for _, r := range stylesheet.Rules {
-			declarations := r.Declarations
-			for _, d := range declarations {
-				s, e := findOffsets(file, d.Line, d.Column, d.Property)
-				out.Refs = append(out.Refs, &graph.Ref{
-					DefUnitType: "URL",
-					DefUnit:     "MDN",
-					DefPath:     mdnDefPath(d.Property),
-					Unit:        u.Name,
-					File:        filepath.ToSlash(currentFile),
-					Start:       uint32(s),
-					End:         uint32(e),
-				})
+		data := string(fileBytes)
+		if isCSSFile(f) {
+			stylesheet, err := cssParser.Parse(data)
+			if err != nil {
```

Comment:
I'd emit a warning and continue instead of returning an error. One failed file is not a reason to abort :)